server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java

```diff
@@ -388,7 +388,7 @@ public Decision canRemain(IndexMetadata indexMetadata, ShardRouting shardRouting
             return decision;
         }
 
-        if (indexMetadata.ignoreDiskWatermarks()) {
+        if (indexMetadata.ignoreDiskWatermarks() || ignoreDiskWatermarksForShard(allocation, shardRouting)) {
```

Comment:
NIT: this already checked the index metadata with `ignoreDiskWatermarks`

Reply:
May be it is worth having an overloaded `ignoreDiskWatermarksForShard` that accepts `indexMetadata` if available already

Reply:
Sure, I pushed 1ffd46749c5619dfd99d3cf512bb55b09aa0229c